Documents/AnalyticsToolDesignDocument.md

```diff
@@ -0,0 +1,48 @@
+# VFS For Git: Analytics Tooling Design Document
+
+This document details the design of and plan for the **VFS4G analytics tool**. The tool aims to help users identify reasons why their enlistment may be slowing down and the `gvfs` commands running slowly. By collecting and reporting statistics, the power to uncover and hopefully fix problems as they arrive moves to the user, empowering them to use the software to its fullest potential.
+
+## Problem
+
+Over the lifetime of an enlistment, sometimes a user's working directory fills up with too many entries. Generally this means having too many placeholders on their disk (have been read and pulled from remote) or too many modified files. This causes unacceptable losses in performance and greatly hinders users as they go about their daily development cycle. Additionally, if a local enlistment has been in use for a long time, the number of loose object may eventually reach the order of several thousand in which case that also contributes to the loss in performance.
+
+## Provided Statistics
+
+### Safe to access
+
+- **Modified paths** - The number of files which have been edited and have been handed over to git
+- **Git index** - The differences between modified paths and the index tracked by git
+- **Placeholders** - A total number of the placeholders being tracked in VFS4G and a list of the placeholders in the system (Note that this contains some upper level directories and thus not all of their children are included in the count without enumeration)
+- **Troublesome programs** - Similar to what is already being logged in the heartbeat, look at which services or programs are responsible for hydrating large amounts of files in the enlistment
+
+### Potentially detrimental to access on a mounted enlistment
+
+- **Placeholder details** - Crawling the file system to enumerate directories of placeholders may unintentionally further hydrate files (IE downloading file sizes), but doing so would allow for greater details such as the percent hydration of folders, the total number of hydrated files, the type of files which are most hydrated, etc.
+- **File system** - Provide enumerated details for files in the placeholder list or the modified path list
+
+## Background Information
+
+- **Modified paths** is the list of files which VFS4G has completely pulled from the remote and whose contents are on disk in their entirety. These files are handed to git via a hook when any git command requiring the list of changed files is called

```

Comment:
A few more details on modified paths.

Modified paths contains:
- Files and directories that have been deleted
- Newly created files/directories (created by the user), including the destination path when a file is renamed.
- File that have been modified by the user (more specifically, opened for write)

Fully hydrated placeholders are not included in modified paths, even though their contents are entirely on disk.

Reply:
I thought fully hydrated placeholders were what files became once they were opened with a write handle. So then that term is for still just placeholders that have some data associated with them from ProjFS?

Reply:
> I thought fully hydrated placeholders were what files became once they were opened with a write handle.

Fully hydrated placeholders are what empty placeholder files become when they're read.

The exact behavior of files is slightly different on Window and Mac:

**Windows**

- Virtual File (*File owned/updated by VFS4G*): File does not exist on disk but is injected into enumeration results by VFS4G.
- Empty Placeholder (*File owned/updated by VFS4G*): File has been opened without write access.  No data has been read from the file.
- Hydrated Placeholder (*File owned/updated by VFS4G*): At least one byte has been read from the file.  If an empty placeholder file is opened for write it will be hydrated as well, but it will immediately transition to the "full file" state.
- Full File (*File owned/updated by Git*): A regular/plain NTFS file.  There is no special ProjFS data on the file.  Newly created files are "full files" as well as placeholder files that were opened for write.
- Tombstone (*File owned/updated by Git*): A special hidden ProjFS file.  Tombstone files suppress virtual files from being displayed during enumeration and they are created when a user deletes a file that VFS4G is currently projecting.

**Mac**
- Empty Placeholder (*File owned/updated by VFS4G*): Created when expanding a directory (Mac specific).  Files remain empty until something attempts to read/write data or attributes.
- Hydrated Placeholder (*File owned/updated by VFS4G*): At least one byte has been read from the file *or* the attributes of the file have been read.  If an empty placeholder file is opened for write it will be hydrated as well, but it will immediately transition to the "full file" state.
- Full File (*File owned/updated by Git*): A regular/plain file.  There is no special ProjFS data on the file.  Newly created files are "full files" as well as placeholder files that were opened for write.

Reply:
Ah okay I see now. I wasn't sure if a "fully hydrated placeholder" was a synonym for "full file" or for "hydrated placeholder" but now I see that it is the latter.